File: mahaam-api-java/src/main/java/mahaam/feat/user/SuggestedEmailsRepo.java
package mahaam.feat.user;

import java.util.List;
import java.util.UUID;

import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import mahaam.feat.user.UserModel.SuggestedEmail;
import mahaam.infra.DB;
import mahaam.infra.Mapper;

public interface SuggestedEmailsRepo {
	UUID create(UUID userId, String email);

	int delete(UUID id);

	List<SuggestedEmail> getMany(UUID userId);

	SuggestedEmail getOne(UUID id);

	int deleteManyByEmail(String email);
}

@ApplicationScoped
class DefaultSuggestedEmailsRepo implements SuggestedEmailsRepo {

	@Inject
	DB db;

	@Override
	public UUID create(UUID userId, String email) {
		String query = """
				INSERT INTO suggested_emails (id, user_id, email, created_at)
				VALUES (:id, :userId, :email, current_timestamp)
				ON CONFLICT (user_id, email) DO NOTHING""";
		UUID id = UUID.randomUUID();
		int updated = db.insert(query, Mapper.of("id", id, "userId", userId, "email", email));
		return updated > 0 ? id : null;

	}

	@Override
	public int delete(UUID id) {
		String query = "DELETE FROM suggested_emails WHERE id = :id";
		return db.delete(query, Mapper.of("id", id));
	}

	@Override
	public List<SuggestedEmail> getMany(UUID userId) {
		String query = """
				SELECT id s_id, user_id s_userId, email s_email, created_at s_createdAt
				FROM suggested_emails s
				WHERE s.user_id = :userId ORDER BY s.created_at DESC""";
		return db.selectList(query, SuggestedEmail.class, Mapper.of("userId", userId));
	}

	@Override
	public SuggestedEmail getOne(UUID id) {
		String query = """
				SELECT id s_id, user_id s_userId, email s_email, created_at s_createdAt
				FROM suggested_emails s
				WHERE s.id = :id""";
		return db.selectOne(query, SuggestedEmail.class, Mapper.of("id", id));
	}

	@Override
	public int deleteManyByEmail(String email) {
		String query = "DELETE FROM suggested_emails WHERE email = :email";
		return db.delete(query, Mapper.of("email", email));
	}
}

File: mahaam-api-cs/Src/Feat/User/SuggestedEmailsRepo.cs

using Mahaam.Infra;

namespace Mahaam.Feat.Users;

public interface ISuggestedEmailsRepo
{
	Guid Create(Guid userId, string email);
	int Delete(Guid id);
	List<SuggestedEmail> GetMany(Guid userId);
	public SuggestedEmail GetOne(Guid id);
	int DeleteManyByEmail(string email);

}

class SuggestedEmailsRepo(IDB db) : ISuggestedEmailsRepo
{
	private readonly IDB _db = db;
	public Guid Create(Guid userId, string email)
	{

		var query = @"INSERT INTO suggested_emails (id, user_id, email, created_at) 
			VALUES (@id, @userId, @email, current_timestamp)
			ON CONFLICT (user_id, email) DO NOTHING";
		var id = Guid.NewGuid();
		var updated = _db.Insert(query, new { id, userId, email });
		return updated > 0 ? id : Guid.Empty;
	}

	public int Delete(Guid id)
	{
		var query = "DELETE FROM suggested_emails WHERE id = @id";
		return _db.Delete(query, new { id });
	}

	public int DeleteManyByEmail(string email)
	{
		var query = "DELETE FROM suggested_emails WHERE email = @email";
		return _db.Delete(query, new { email });
	}

	public List<SuggestedEmail> GetMany(Guid userId)
	{
		var query = @"SELECT id, user_id, email, created_at
			FROM suggested_emails WHERE user_id = @userId order by created_at desc";
		return _db.SelectMany<SuggestedEmail>(query, new { userId });
	}

	public SuggestedEmail GetOne(Guid id)
	{
		var query = @"SELECT id, user_id, email, created_at FROM suggested_emails WHERE id = @id";
		return _db.SelectOne<SuggestedEmail>(query, new { id });
	}
}